Enable regex for directory directives

pbernery opened the pull request:
The code was already ready for that but escaped the potential regexp declared in the reek configuration, actually disabling regexp.

This commit fixes this by removing the call to escape.

This allows this kind of declaration in the configuration file:

\# .reek.yml
```yaml
---
[...]
  "engines/.*/test/.*":
    InstanceVariableAssumption:
      enabled: false
    UncommunicativeVariableName:
      enabled: false
```

If this PR is accepted, I'll update the documentation to show this _feature_.

pbernery commented:
Note that with this solution, if there is this directive "engines/.*/test/.*", it only trigger for subfolders of  "engines/.*/test/", not for files in "engines/.*/test/", meaning the directive may have to be duplicated.

I believe this part of the code should use a Dir.glob like syntax, which is commonly used in Ruby configuration files. The example of this PR would look like this in that case:

```yaml
---
[...]
  "engines/**/test/**":
[...]
```

What do you think?

troessner commented:
>> I believe this part of the code should use a Dir.glob like syntax, which is commonly used in Ruby configuration files. The example of this PR would look like this in that case:

I like the idea. @mvz wdyt?

mvz commented:
Yes, using glob-like syntax would definitely improve this :+1:.

We are currently matching by substring, but it seems that is an accident. It is also undocumented :man_shrugging:, so we can safely make this improvement.

troessner commented:
Sounds good.  @pbernery are you up for updating your PR accordingly?

pbernery commented:
Sure, I'll have a look into that.

pbernery commented:
I've spent some time trying to implement something with `Dir.glob`.

There will be an incompatibility with the existing configuration: currently, a directory directive like "foo/bar/baz" sets a configuration for this directory and all files under it. 
With `Dir.glob`, "foo/bar/alpha" will match exactly this directory but not its subdirectories. To include subdirectories, the configuration should be "foo/bar/alpha/**".

Do you think we can ask user to update their configuration if this feature is integrated?
I also thought about adding a new configuration key, which would be something like "file_directives": using `Dir.glob`, we can let user specify rules for file patterns instead of just directories. For instance, something like "test/**/*_test.rb".

troessner commented:
>> Do you think we can ask user to update their configuration if this feature is integrated?

That's not really an option but we can do a new major release for this given that the feature seems quite useful.

mvz commented:
We can simply assume trailing `**/*` for current values, I think.

Also, don't we just blanket stop descending into directories once they're found to be ignored?

pbernery commented:
> I think I'd rather not have a new key but instead have a major release, a short migration guide (can be a paragraph and an example, nothing fancy) and some updated docs.

I like this idea too.

> We can simply assume trailing `**/*` for current values, I think.

Yes, and no. Yes if you don't want to support exact match, but this is not an expected behaviour when having configuration based on `Dir.glob`, this could lead to some incomprehension from users I believe.

> Also, don't we just blanket stop descending into directories once they're found to be ignored?

I am not sure I can do that easily with an implementation based on `Dir.glob` as complex patterns may be given.

mvz commented:
I would prefer to implicitly add `**/*` at the end over having a breaking change at this time.

The most important part is fleshing out how this feature should behave with some specs, and how that differs from the current behavior. There is currently a bit of logic to find the 'best' match, for example, and we should think about how that will behave in any new system.

To sum up, I like the idea of allowing globbing, but we need to think a bit more about how that should work exactly.

pbernery commented:
Alright.

So I can start with a first implementation with implicit `**/*` and we can then discuss from that. This is actually what I did at first, I should have something rapidly.

I'll try to push something this week.

pbernery commented:
Sorry for the delay, here is a proposition of a configuration that is compatible with the current one, and that supports `Dir.glob` like patterns.

troessner commented:
This looks good to me :)
@mvz thoughts?
@pbernery let's add some documentation (to this PR) and then make this ready to merge unless @mvz has comments.

mvz commented:
I think this is definitely the way to go at this time :+1:.

A couple of things I think we should look at:

* Any `.` and perhaps some other characters will need to be escaped.
* `**` should indeed expand to `.*`, but `*` should probably become `[^/]*` or similar.

pbernery commented:
I updated the documentation and added a gsub statement to escape `.`.

@mvz, there is already a rule to replace `*` by `[^/]*` if I am not mistaken. Or did I miss something?

pbernery commented:
I noticed the CI didn't pass because of an alignment issue. I fixed that.

Do you think this PR is good to merge or is there something missing?

troessner commented:
I'd be good with merge-squashing it like it is. @mvz you good as well?

troessner commented:
Merged! Thank you @pbernery, I'll release it right away.

troessner commented:
And released it is as 5.4.0 ;)
